### PR TITLE
starlark: tweak how UnpackArgs enforces optional args

### PR DESCRIPTION
**Pull request by nicks**
Fixes https://github.com/google/starlark-go/issues/544

### CODE REVIEW

File: starlark/unpack.go [hunk 1]
```diff
@@ -44,23 +44,23 @@ type Unpacker interface {
 //
 // Examples:
 //
-//      var (
-//          a Value
-//          b = MakeInt(42)
-//          c Value = starlark.None
-//      )
+//	var (
```
Comment (nicks) [inline starlark/unpack.go:47]: fyi -- these indentation changes come from go fmt in Go 1.22. I can separate them out into a separate pr if that's easier.

File: starlark/eval_test.go [hunk 2]
```diff
@@ -1109,3 +1109,44 @@ f(1)
...
+	kwargs := []starlark.Tuple{
+		{starlark.String("x"), starlark.MakeInt(1)},
+		{starlark.String("y"), starlark.MakeInt(2)},
+	}
+	var x, y, z int
+	if err := starlark.UnpackArgs("unpack", nil, kwargs,
+		"x", &x, "y?", &y, "z", &z); err != nil {
+		t.Errorf("UnpackArgs failed: %v", err)
+	}
+	if x != 1 {
```
Comment (adonovan) [inline starlark/eval_test.go:1124]: You can achieve higher vertical density (and, in this case, readability) in both the code and the test log by combining all three in one line:

got := fmt.Sprint("x=%d y=%d z=%d", x, y, z)
want := "x=1 y=2 z=0"
...assert...

File: starlark/unpack.go (hunk 1 above)
Comment (adonovan) [reply starlark/unpack.go:47]: Don't worry, it's fine.

Comment (adonovan) [reply starlark/unpack.go:47]: While you're here, would you mind fixing these preexisting things?

L28: s/integer types uses/integer types use/
L36:  s/where or not/whether or not/

s/interface{}/any/g throughout this function

thanks

File: starlark/unpack.go [hunk 3]
```diff
@@ -164,12 +163,15 @@ kwloop:
 	}
 
 	// Check that all non-optional parameters are defined.
-	// (We needn't check the first len(args).)
-	for i := len(args); i < nparams; i++ {
+	for i := 0; i < nparams; i++ {
 		name := pairs[2*i].(string)
 		if strings.HasSuffix(name, "?") {
 			break // optional
```
Comment (adonovan) [inline starlark/unpack.go:169]: The doc comment says "If a parameter is marked optional, then all following parameters are implicitly optional whether or not they are marked." In hindsight we should probably have disallowed non-optionals after optionals, but that would be a breaking change now.

**Review by adonovan — COMMENTED** [review]
Thanks for the fix!

File: starlark/unpack.go (hunk 3 above)
Comment (nicks) [reply starlark/unpack.go:169]: agreed

**Review by adonovan — APPROVED** [review]
Hi Nick, thanks for sending this fix. LGTM.

**Pull request merged**